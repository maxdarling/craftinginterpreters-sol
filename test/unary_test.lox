// Test unary operators (-, !)
print "=== Unary Operators Tests ===";

// Negation operator (-)
print -5; // -5
print -(-5); // 5
print -0; // -0 (which is 0)
print -1.5; // -1.5
print -(-1.5); // 1.5

// Negation with variables
var x = 10;
print -x; // -10
print x; // 10 (x should be unchanged)

// Negation with expressions
print -(1 + 2); // -3
print -(5 * 2); // -10
print -(10 / 2); // -5.0

// Logical NOT operator (!)
print !true; // false
print !false; // true
print !!true; // true
print !!false; // false

// NOT with truthy/falsy values (only nil and false are falsy)
print !nil; // true (nil is falsy)
print !0; // false (0 is truthy)
print !""; // false (empty string is truthy)
print !1; // false (1 is truthy)
print !"hello"; // false (non-empty string is truthy)

// NOT with expressions
print !(1 == 1); // false
print !(1 != 1); // true
print !(5 > 3); // false
print !(5 < 3); // true

// Complex unary expressions
print !!(1 + 1); // true (double negation of truthy value)
print !!nil; // false (double negation of falsy value)
print -(-(-5)); // -5 (triple negation)

// Mixed unary and binary operators
print -5 + 3; // -2
print !true and false; // false
print !false or true; // true
print -(2 * 3) + 1; // -5

// Unary with function calls
fun getNumber() {
  return 42;
}

print -getNumber(); // -42
print !getNumber(); // false (42 is truthy)

print "Unary tests completed";
