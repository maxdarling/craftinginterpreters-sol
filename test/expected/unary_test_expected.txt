=== Unary Operators Tests ===
-5
5
-0
-1.5
1.5
-10
10
-3
-10
-5
false
true
true
false
true
false
false
false
false
false
true
false
true
true
false
-5
-2
false
true
-5
-42
false
Unary tests completed
